### PR TITLE
proxmox_template download appliance templates

### PR DESCRIPTION
**Pull request by grebsai**
##### SUMMARY
Download proxmox applicance templates

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
proxmox_template

##### ADDITIONAL INFORMATION
Now it is possible to download Proxmox appliance templates.
Proxmox API https://pve.proxmox.com/pve-docs/api-viewer/index.html#/nodes/{node}/aplinfo

```
- name: Download proxmox appliance container template
  community.general.proxmox_template:
    node: uk-mc02
    api_user: root@pam
    api_password: 1q2w3e
    api_host: node1
    storage: local
    content_type: vztmpl
    template: ubuntu-20.04-standard_20.04-1_amd64.tar.gz
```

### CODE REVIEW

**Comment by grebsai** [comment]
ready_for_review

File: changelogs/fragments/proxmox_template-appliance-download.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - proxmox_template - download proxmox applicance templates (pveam)
```
Comment (felixfontein) [inline changelogs/fragments/proxmox_template-appliance-download.yml:3]:
```suggestion
  - proxmox_template - download proxmox applicance templates (pveam) (https://github.com/ansible-collections/community.general/pull/1046).
```

File: plugins/modules/cloud/misc/proxmox_template.py [hunk 2]
```diff
@@ -48,7 +48,8 @@
   template:
     description:
       - the template name
-      - required only for states C(absent), C(info)
+      - required for state C(absent) to delete a template
+      - required for state C(present) to download an appliance container template (pveam)
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_template.py:52]:
```suggestion
      - Required for state C(absent) to delete a template.
      - Required for state C(present) to download an appliance container template (pveam).
```
(Let's start doing at least the modified ones correctly :) )

**Comment by felixfontein** [comment]
@Aversiste @UnderGreen @joshainglis @karmab Dear maintainers, please take a look at this PR :)

File: plugins/modules/cloud/misc/proxmox_template.py [hunk 3]
```diff
@@ -216,6 +240,19 @@ def main():
             content_type = module.params['content_type']
             src = module.params['src']
 
+            # download appliance template
+            if content_type == 'vztmpl' and not src:
+                template = module.params['template']
+
+                if not template:
+                    module.fail_json(msg='template param to downloading appliance template is mandatory')
```
Comment (Aversiste) [inline plugins/modules/cloud/misc/proxmox_template.py:248]: English is not my mother tongue but shouldn't it be “for downloading”?

File: plugins/modules/cloud/misc/proxmox_template.py [hunk 4]
```diff
@@ -216,6 +240,19 @@ def main():
...
+
+                if get_template(proxmox, node, storage, content_type, template) and not module.params['force']:
+                    module.exit_json(changed=False, msg='template with volid=%s:%s/%s is already exists' % (storage, content_type, template))
```
Comment (Aversiste) [inline plugins/modules/cloud/misc/proxmox_template.py:251]:
```suggestion
                    module.exit_json(changed=False, msg='template with volid=%s:%s/%s already exists' % (storage, content_type, template))
```

**Review by Aversiste — COMMENTED** [review]
In my environment I clone stopped VM rather than using templates (I know, it's not the best) so I don't have much testing to provide.

File: plugins/modules/cloud/misc/proxmox_template.py (hunk 3 above)
Comment (grebsai) [reply plugins/modules/cloud/misc/proxmox_template.py:248]: you're right

**Comment by grebsai** [comment]
ready_for_review

File: plugins/modules/cloud/misc/proxmox_template.py [hunk 5]
```diff
@@ -140,21 +151,34 @@ def get_template(proxmox, node, storage, content_type, template):
...
 
-def upload_template(module, proxmox, api_host, node, storage, content_type, realpath, timeout):
-    taskid = proxmox.nodes(node).storage(storage).upload.post(content=content_type, filename=open(realpath, 'rb'))
+def task_status(module, proxmox, node, taskid, timeout):
+    """
+    Check the task status and wait until the task is completed or the timeout is reached.
+    """
     while timeout:
-        task_status = proxmox.nodes(api_host.split('.')[0]).tasks(taskid).status.get()
+        module.log("timeout")
```
Comment (Aversiste) [inline plugins/modules/cloud/misc/proxmox_template.py:159]: Looks like a debug leftover :)

**Review by Aversiste — COMMENTED** [review]
I do not use this module but ok for me as it looks good.

File: plugins/modules/cloud/misc/proxmox_template.py (hunk 5 above)
Comment (grebsai) [reply plugins/modules/cloud/misc/proxmox_template.py]: ups, thanks. removed it

**Comment by felixfontein** [comment]
If nobody complains, I'll merge this in a few days.

**Comment by felixfontein** [comment]
@grebsai thanks for implementing this!
@Aversiste thanks for reviewing this!

**Pull request merged**